src/V3Delayed.cpp

```diff
@@ -118,307 +125,297 @@ class DelayedVisitor final : public VNVisitor {
+        // Running node pointer
+        AstNodeExpr* nodep = components.refp;
+        // Apply AstArraySels
+        for (AstNodeExpr* const idxp : components.arrIdxps) {
+            nodep = new AstArraySel{flp, nodep, idxp};
         }
-        if (dimselp) varrefp = VN_AS(dimselp, VarRef);
-        UASSERT_OBJ(varrefp, nodep, "No var underneath arraysels");
-        UASSERT_OBJ(varrefp->varScopep(), varrefp, "Var didn't get varscoped in V3Scope.cpp");
-        varrefp->unlinkFrBack();
-        const AstVar* const oldvarp = varrefp->varp();
-        const int modVecNum = m_scopeVecMap[varrefp->varScopep()]++;
-        //
-        std::deque<AstNodeExpr*> dimreadps;  // Read value for each dimension of assignment
-        for (unsigned dimension = 0; dimension < dimvalp.size(); dimension++) {
-            AstNodeExpr* const dimp = dimvalp[dimension];
-            if (VN_IS(dimp, Const)) {  // bit = const, can just use it
-                dimreadps.push_front(dimp);
-            } else {
-                const string bitvarname = string{"__Vdlyvdim"} + cvtToStr(dimension) + "__"
-                                          + oldvarp->shortName() + "__v" + cvtToStr(modVecNum);
-                AstVarScope* const bitvscp
-                    = createVarSc(varrefp->varScopep(), bitvarname, dimp->width(), nullptr);
-                AstAssign* const bitassignp = new AstAssign{
-                    nodep->fileline(), new AstVarRef{nodep->fileline(), bitvscp, VAccess::WRITE},
-                    dimp};
-                nodep->addNextHere(bitassignp);
-                dimreadps.push_front(new AstVarRef{nodep->fileline(), bitvscp, VAccess::READ});
-            }
+        // Apply AstSel, if any
+        if (components.selLsbp) {
+            nodep = new AstSel{flp, nodep, components.selLsbp, components.selWidthp};
         }
-        //
-        //=== Bitselect: __Vdlyvlsb__
-        AstNodeExpr* bitreadp = nullptr;  // Code to read Vdlyvlsb
-        if (bitselp) {
-            AstNodeExpr* const lsbvaluep = bitselp->lsbp()->unlinkFrBack();
-            if (VN_IS(lsbvaluep, Const)) {
-                // vlsb = constant, can just push constant into where we use it
-                bitreadp = lsbvaluep;
-            } else {
-                const string bitvarname
-                    = string{"__Vdlyvlsb__"} + oldvarp->shortName() + "__v" + cvtToStr(modVecNum);
-                AstVarScope* const bitvscp
-                    = createVarSc(varrefp->varScopep(), bitvarname, lsbvaluep->width(), nullptr);
-                AstAssign* const bitassignp = new AstAssign{
-                    nodep->fileline(), new AstVarRef{nodep->fileline(), bitvscp, VAccess::WRITE},
-                    lsbvaluep};
-                nodep->addNextHere(bitassignp);
-                bitreadp = new AstVarRef{nodep->fileline(), bitvscp, VAccess::READ};
-            }
+        // Done
+        return nodep;
+    }
+
+    void createDlyOnSet(AstAssignDly* nodep) {
```

Comment:
Sizewise this is mostly as it was before. It becomes smaller in the follow up patches were it is chopped up a bit.

> Can you make functions that take a ref rather than lambdas?

Well, of course I *can*, but I really don't want to :) For me it makes the code more opaque to have to explicitly pass non-varying parameter consistently every time. Also, the call sites become longer and the formatter breaks them up in a hard to read way. The small lambdas with clear defined behaviour show which parts wary (those you pass in), and which part doesn't (those that are captured) per cal site.

Reply:
Ok, fine, feel free to merge if ready.